flink-runtime-web/src/test/java/org/apache/flink/runtime/webmonitor/history/HistoryServerTest.java

```diff
@@ -117,45 +124,145 @@ public void testHistoryServerIntegration() throws Exception {
+							break;
+					}
+				});
 
-		HistoryServer hs = new HistoryServer(historyServerConfig, numExpectedArchivedJobs);
 		try {
 			hs.start();
+			// check that expected count of archives exists
 			String baseUrl = "http://localhost:" + hs.getWebPort();
 			numExpectedArchivedJobs.await(10L, TimeUnit.SECONDS);
 
 			ObjectMapper mapper = new ObjectMapper();
-			String response = getFromHTTP(baseUrl + JobsOverviewHeaders.URL);
-			MultipleJobsDetails overview = mapper.readValue(response, MultipleJobsDetails.class);
+			Assert.assertEquals(numJobs, getJobsOverview(baseUrl, mapper).getJobs().size());
+
+			String jobIDToDelete = jobIDs.get(0);
+			verifyArchiveFilesExistence(jobIDToDelete, true);
+
+			// delete one archive from jm
+			URI resolve = jmDirectory.toURI().resolve(jobIDToDelete);
+			new File(resolve).delete();
+
+			numExpectedExpiredJobs.await(10L, TimeUnit.SECONDS);
+
+			// check that archive presence in hs
+			Assert.assertEquals(numJobs - numExpiredJobs, getJobsOverview(baseUrl, mapper).getJobs().size());
+			verifyArchiveFilesExistence(jobIDToDelete, !cleanupExpiredJobs);
 
-			Assert.assertEquals(numJobs + 1, overview.getJobs().size());
 		} finally {
 			hs.stop();
 		}
 	}
 
-	private static void runJob() throws Exception {
+	private void verifyArchiveFilesExistence(String jobID, boolean expectArchiveFilesExist) {
+		List<File> archiveFiles = Arrays.asList(
+			new File(hsDirectory.toURI().resolve(String.format("overviews/%s.json", jobID))),
+			new File(hsDirectory.toURI().resolve(String.format("jobs/%s", jobID))),
+			new File(hsDirectory.toURI().resolve(String.format("jobs/%s.json", jobID))));
+
+		archiveFiles.forEach(
+			archiveFile -> Assert.assertEquals(
+				String.format("Existence of file or directory %s", archiveFile.getPath()),
+				archiveFile.exists(), expectArchiveFilesExist));
+	}
+
+	private void waitForArchivesCreation(int numJobs) throws InterruptedException {
+		// the job is archived asynchronously after env.execute() returns
+		File[] archives = jmDirectory.listFiles();
+		while (archives == null || archives.length != numJobs) {
+			Thread.sleep(50);
+			archives = jmDirectory.listFiles();
+		}
+	}
+
+	private Configuration createTestConfiguration(boolean cleanupExpiredJobs) {
+		Configuration historyServerConfig = new Configuration();
+		historyServerConfig.setString(HistoryServerOptions.HISTORY_SERVER_ARCHIVE_DIRS, jmDirectory.toURI().toString());
+		historyServerConfig.setString(HistoryServerOptions.HISTORY_SERVER_WEB_DIR, hsDirectory.getAbsolutePath());
+		historyServerConfig.setLong(HistoryServerOptions.HISTORY_SERVER_ARCHIVE_REFRESH_INTERVAL, 100L);
+
+		historyServerConfig.setBoolean(HistoryServerOptions.HISTORY_SERVER_CLEANUP_EXPIRED_JOBS, cleanupExpiredJobs);
+
+		historyServerConfig.setInteger(HistoryServerOptions.HISTORY_SERVER_WEB_PORT, 0);
+		return historyServerConfig;
+	}
+
+	private MultipleJobsDetails getJobsOverview(String baseUrl, ObjectMapper mapper) throws Exception {
```

Comment:
you can simplify this method by having a static final ObjectMapper field
this method can also be static.

Reply:
Good point. Constant created and method refactored. Thanks